integration/doc/oasis/charge_station_based_routing.md

```diff
@@ -0,0 +1,88 @@
+                       -> convert csv to json 
+                                 -> decoding json
+```
+More information [#237](https://github.com/Telenav/osrm-backend/issues/237) 
+
+
+
+### JSON Input
+Electricity charge stations, expect input in JSON format follow OSM's tag definition about [node](https://wiki.openstreetmap.org/wiki/Node) and [charging_station](https://wiki.openstreetmap.org/wiki/Tag:amenity%3Dcharging_station), example:
+```json
+{
+   "id": 12345,
+   "lat": 51.5173639,
+   "lon": -0.140043,
+   "amenity": "charging_station",
+   "operator": "NOFT",
+   "fee": "yes",
+   "capacity":4,
+}
+```
+- Later could implement more input connector: pbf, csv, etc.
+
+### Pre-Processing
+- For each charge station point, calculate its google s2 cellid
+   + For each point, it could have cellid in [0, 30] levels.  We just pick a range from them, such as [10, 20].
+   + The range used for pre-processing will be the same as query
+- For each charge station point, take which as center and construct a circle whose radius is a hard code max drive range value(500km), query for all cells it could touch, only record the one contains charge stations
+  + Make sure such function could work: give a cell id(not station id) could retrieve all cells it could reached by.  We also need to distinguish the stations could be reached by charging how much of energy
+  + Make sure exact distance between charge stations has been recorded in db
+  + For production, circle should be replaced by `reach range`(or `isoline`, consider elevation, red/lights, energy pattern, etc)
+- More infor about google S2
+  + [How S2 cell is generated](https://github.com/CodeBear801/tech_summary/blob/master/tech-summary/navigation/spatial_index/google_s2_cell.md)
+  + [How hilbert curve works](https://github.com/CodeBear801/tech_summary/blob/master/tech-summary/navigation/spatial_index/hilbert_curve.md)
+  + [#236](https://github.com/Telenav/osrm-backend/issues/236)
+
+### Query
+- Given two points of orig/dest
+- Check whether orig/dest is reachable by current energy
+- If not start two strategy together: `charge station based routing` and `search along route`
+- logic of `charge station routing`
+   + For input, need using its current energy to filter all charge station it could reached
+   + For Dest, check all charge station needed to reach destination
+        *  Check whether there is overlap between those two, if yes, then go to next step
+        * For all reachable stations from start, generate a reachable cellids by combine each charge station's 1-time-charge reachable cells result, see whether there is any overlap with dest's stations(or cellids contain those stations), if yes, go to next step
+        * For each reachable charge station of start, they have their unique value of `arrival energy`, this value need to be considered during calculating `charge times`, but the exist status of each charge station is fixed: such as charge to 60% of energy, 80% of energy, etc.
```

Comment:
What does the meaning of `exist status`?

Reply:
* For each reachable charge station of start, they have their unique value of `arrival energy`, this value need to be considered during calculating `charge times`, but the `exist status` of each charge station is fixed: such as charge to 60% of energy, 80% of energy, etc.

![image](https://user-images.githubusercontent.com/16873751/88236950-09eac280-cc33-11ea-9698-76adc9ce6087.png)

Reply:
So it's `target charging energy` when charge on this station, right? The name `exist status` does not make sense, may be it's a typo that you want to use `exit status`?

Reply:
You are right, using `target energy` is better, which is more align with code

```go

// definition
// State contains charging related information
type State struct {
	Energy float64
}

// CreateChargingStates generates charging status
func (f *chargeStrategy) CreateChargingStates() []State {
}


// client
for _, targetState := range g.strategy.CreateChargingStates() {

}
```

Reply:
I changed the description to `target status` I think which makes more sense.